clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp

```diff
@@ -264,44 +271,86 @@ Expected<SmallVector<std::string>> getSYCLDeviceLibs(const ArgList &Args) {
-    return LLVMLinkPath.takeError();
-
+/// Following tasks are performed:
+/// 1. Link all SYCL device bitcode images into one image. Device linking is
+/// performed using the linkModules API.
+/// 2. Gather all SYCL device library bitcode images.
+/// 3. Link all the images gathered in Step 2 with the output of Step 1 using
+/// linkModules API. LinkOnlyNeeded flag is used.
+Expected<StringRef> linkDeviceCode(ArrayRef<OffloadFile> BitcodeInputFiles,
+                                   const ArgList &Args, LLVMContext &C) {
+  llvm::TimeTraceScope TimeScope("SYCL Link Device Code");
+  // Gather SYCL device library files.
+  auto SYCLDeviceLibFilenames = getSYCLDeviceLibs(Args);
+  if (!SYCLDeviceLibFilenames)
+    return SYCLDeviceLibFilenames.takeError();
   // Create a new file to write the linked device file to.
-  auto OutFileOrErr =
+  auto BitcodeOutput =
       createTempFile(Args, sys::path::filename(OutputFile), "bc");
-  if (!OutFileOrErr)
-    return OutFileOrErr.takeError();
-
-  SmallVector<StringRef, 8> CmdArgs;
-  CmdArgs.push_back(*LLVMLinkPath);
-  CmdArgs.push_back("-only-needed");
-  CmdArgs.push_back(InputFile);
-  for (auto &File : *SYCLDeviceLibFiles)
-    CmdArgs.push_back(File);
-  CmdArgs.push_back("-o");
-  CmdArgs.push_back(*OutFileOrErr);
-  CmdArgs.push_back("--suppress-warnings");
-  if (Error Err = executeCommands(*LLVMLinkPath, CmdArgs))
-    return std::move(Err);
-  return *OutFileOrErr;
+  if (!BitcodeOutput)
+    return BitcodeOutput.takeError();
+  int FD = -1;
+  if (std::error_code EC = sys::fs::openFileForWrite(*BitcodeOutput, FD))
+    return errorCodeToError(EC);
+  if (DryRun || Verbose) {
+    std::for_each((*SYCLDeviceLibFilenames).begin(),
+                  (*SYCLDeviceLibFilenames).end(),
+                  [](std::string &Filename) { Filename += "; "; });
+    std::string LibInputs =
+        std::accumulate((*SYCLDeviceLibFilenames).begin(),
+                        (*SYCLDeviceLibFilenames).end(), std::string(""));
+
+    auto FilenamesOrErr = getObjectsFromArgs(Args);
+    if (!FilenamesOrErr)
+      reportError(FilenamesOrErr.takeError());
+    std::for_each((*FilenamesOrErr).begin(), (*FilenamesOrErr).end(),
+                  [](std::string &Filename) { Filename += "; "; });
+    std::string Inputs = std::accumulate(
+        (*FilenamesOrErr).begin(), (*FilenamesOrErr).end(), std::string(""));
+
+    errs() << formatv(
+        "sycl-device-link: input: {0} libfiles: {1} output: {2}\n", Inputs,
+        LibInputs, *BitcodeOutput);
+    if (DryRun)
+      return *BitcodeOutput;
+  }
+  SMDiagnostic Err;
+  assert(BitcodeInputFiles.size() && "No inputs to link");
+  auto Output = std::make_unique<Module>("sycl-device-link", C);
+  Linker L(*Output);
+  // Link SYCL device input files.
+  for (auto &File : BitcodeInputFiles) {
+    auto TheImage = File.getBinary()->getImage();
+    std::unique_ptr<Module> ModOrErr =
+        parseIR(MemoryBufferRef(TheImage, ""), Err, C);
```

Comment:
Can we use `getLazyIRModule` instead of `parseIR`?

Reply:
I tried that and ran into issues related to unique pointers. I will add a TODO and address this later.

Thanks